Book 1 | Fig 3.9

VibhutiBansal-11 opened the pull request:
## Description
Added gauss_infer_2d.ipynb for [gauss_infer_2d](https://github.com/probml/pyprobml/blob/master/notebooks/book1/03/gauss_infer_2d.ipynb) 
<!-- Please refer to https://github.com/probml/pyprobml/blob/master/CONTRIBUTING.md before opening this PR -->

### Figure Number : Book 1 - 3.9

<!-- Mention figure number with corresponding book (book1 or book2). For example: book1 - 4.18 -->

### Figures

<!-- Create a public gist with rendered figures and add a link below. 
Refer to this template for example: https://gist.github.com/patel-zeel/28a101009f5e938082b6d93101bf1fd7 -->
[gist](https://gist.github.com/VibhutiBansal-11/0ca5ecd1b36e2adb9d2b43dafbb4b126)
### Issue 

<!-- Link the issue you are solving -->

<!-- If the issue is from this repo, include directly with issue number -->
<!-- For example: 

#12

-->

<!-- If the issue is from another repo, you can include it using the regular linking mechanism -->
<!-- For example:

-->


Pyprobml Issue (https://github.com/probml/pyprobml/issues/739) 
## Steps

<!-- If the task is complicated, you can specify each step using checkboxes so that everyone can understand what you have done and what you are going to complete. -->
<!-- For example:

- [ ] Step1
- [ ] Step2 

-->
 - [x] Changed the name of the variables/functions to PEP 8 style
  - [x] Added code to generate two figures as given in the book and Latexified them.
  - [x] Converted code to JAX
  - [x] Added interactive demo to change random_state, number of sample data points
## Checklist:

- [x] Performed a self-review of the code
- [x] Tested on Google Colab.

## Potential problems/Important remarks

<!-- If you have any important remarks for the reviewers to take a closer look, you can write them in detail here.  -->
- I opened a new issue and closed the previous one and I created a branch, from now on will make changes per figure in the same issue

karm-patel commented:
## Figure related comments
Hi @VibhutiBansal-11, the new figure looks excellent! I have some comments
1. @patel-zeel gave me this suggestion: Look at this below figure 
![image](https://user-images.githubusercontent.com/59387624/167080790-af281e30-3c04-4e60-9257-5d304888f1d2.png)
 - If we use a plot like this then we don't need to use that color bar since the value is written in the plot itself, of course, if contour lines overlap then we can decrease the number of lines by using `levels` argument I guess.
 - Source of this book (page 51): https://github.com/rougier/scientific-visualization-book/blob/master/pdf/book.pdf
 
2. Try to plot the above figure, if that looks good we can go ahead with that. Otherwise in the current version, I have the following comments
    1. Ideally color bar height should be the same as the plot height - Zeel
    2. Observe the figure, It does not look like square, you can use 
     `plt.gca().set_aspect(equal)` [Source:](plt.gca().set_aspect(%22equal%22)) ~ Zeel
    3. In 2nd subplot, I think title is too much wider, what do you think? We can remove word "contour plot" from the title, since it's visible.
    4.(I'll add more comments)

VibhutiBansal-11 commented:
Hi @karm-patel
b  (fontsize 12)      |  c (fontsize 6)
:-------------------------:|:-------------------------:
 ![b](https://user-images.githubusercontent.com/52831429/167266306-b537c0b9-2e87-4df0-b29c-9cebcd6046f5.png) |![c](https://user-images.githubusercontent.com/52831429/167266307-8eccf766-ff79-4c60-8ae9-3500f350ea36.png)
 
 I made these figures but the font size have been changed as the space between two contour levels was less than font size , so the image looked very weird with same larger font

Yes, I think we could keep the title of second subplot as prior only i guess.

VibhutiBansal-11 commented:
b  (fontsize 12)      |  c (fontsize 6)
:-------------------------:|:-------------------------:
![b_1](https://user-images.githubusercontent.com/52831429/167269495-8c803a1d-519e-4e3c-bba1-74d32feb1545.png)|![c_1](https://user-images.githubusercontent.com/52831429/167269403-cdeff235-fd1a-4754-9935-2d58f93b894d.png)

karm-patel commented:
Hi @VibhutiBansal-11, yes so this trick (writing the value in plot) is not working in 3rd subplot, so I guess we can stick the figure which you have created, I am adding comments here.
1. Ideally color bar height should be the same as the plot height - Zeel
2. In the first subplot, we should mention a cross symbol in the legend. Also, can x_label and y_label be written as "y1" and "y2"? we should refer to the text/caption.
3. In the last 2 plots, we can keep x_label and y_label as "z1" and "z2", since the book mentioned the random variable as 'z', can you please cross-check this with book text since I'm not sure.

karm-patel commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,709 @@
+{
```
<p>Please add meaningfull heading in 1st markdown cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

karm-patel replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    latexify(width_scale_factor=2, fig_height=2)
</pre><p>width_scale_factor shoulde be 3 since we have 3 figures.</p><p>fig_height should be 1.5 since  I have checked in latex source code</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

karm-patel replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    def make_graph(x,muTrue,savename,title,fig=None,ax=None):
</pre><p>Function can be renamed as plot_data() ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

karm-patel commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,709 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #4.        print("probml_utils module not found")
</pre><p>We should follow the same notation, like we are generally not writing print() here, instead it should be something like </p><p>try:</p><p>&nbsp;&nbsp;from probml_utils import savefig, latexify</p><p>except ModuleNotFoundError :</p><p>&nbsp;&nbsp;%pip install ....</p><p>   from probml_utils import savefig, latexify</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

VibhutiBansal-11 replied:
<p>Okay noted.</p>

karm-patel commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,709 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    def make_graph2(x1,x2,p,savename,title,fig=None,ax=None):
</pre><p>Function can be renamed as make_contour_graph() ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

VibhutiBansal-11 replied:
<p>Okay I will change according to the above mentioned comments, but I was confused about the x,y labels for the subplots.</p>

karm-patel replied:
You can update label as following
|fig | xlabel | ylabel |
|-- | -- | -- |
| a) | y1 | y1 |
| b) | z1 | z2 |
| c) | z1 | z2 |

let me know what do you think?

karm-patel commented:
I have added some comments in code, once our figure is finalized I'll take a detail look in code.

VibhutiBansal-11 commented:
I have made changes as per the above suggestions
- [x]  Made the sqaure plots and changed its latexify parameters to 3,1.5
- [x] Changed the x,y labels
- [x] Added colorbar of equal height
- [x] renamed the functions
- [x] changed the way of importing probml utils
Given below is an updated gist

[Gist](https://gist.github.com/VibhutiBansal-11/a26ce056f3bb9c433baa9ed61bdafbf4)

karm-patel commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,757 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #11.      ax.plot(muTrue[0], muTrue[1], "x", linewidth=4, markerfacecolor="black", color="k",label='z')
</pre><p>What it is need of to add argument <code>linewidth=4? </code> Because now linewidth=4 will be applied in latexified and default (notebook) version.  if you observe latexified figure, the plot points look too much large than usual. Hence, we should not change default setting in latexified figute until it is required. </p><p> </p><p>You might think that, if you will not add <code>linewidth=4</code> then dots in default figure in notebook  looks very small, in that case you can use method <code>is_latexify_enabled()</code> from <code>probml_utils</code>.</p><p> </p><p>You can write like below:</p><pre class="ql-syntax" spellcheck="false">LINE_WIDTH = None if is_latexify_enabled() else 4

ax.plot(muTrue[0], muTrue[1], "x", linewidth=LINE_WIDTH, markerfacecolor="black", color="k",label='z')
</pre><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

karm-patel commented:
## Figure related comments
1. Scatter size First subplot looks big, I have mentioned this comment in code also'
2. Legend symbol size and legend text size does not match, mentioned this comment in the code also, Ideally we should follow latexify default setting.

karm-patel commented:
Also, once you resolve my comments please click on `resolve conversation`.

VibhutiBansal-11 commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,757 @@
+{
```
<p>Hi, this was added as after changing latexify width size and height factors image in latexified version was too small but the legend text was covering the whole image that's why I tried these work arounds, but I understand your point. Shall update the changes.</p>

VibhutiBansal-11 commented:
@karm-patel I have made the necessary changes, but in plot 1 if we keep font size above 7 then the legend doesn't fit the latexified image. So, kept the font size at 7.

karm-patel commented:
Great! Can you please paste the updated figure here?

karm-patel commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,752 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #4.      ax.plot(x[:, 0], x[:, 1], "o",  markerfacecolor="b",label='$y_n$')
</pre><p>label = "$samples (y_n)$" make sense?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

karm-patel commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,752 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #10.      ax.plot(muTrue[0], muTrue[1], "x",  markerfacecolor="black", color="k",label='z')
</pre><p>label = '$mean (z)$' make sense?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

VibhutiBansal-11 replied:
Hi @karm-patel , actually the figure is too small to add big labels , we would have to reduce the label font size if we increase the text length

karm-patel replied:
I see, can you please paste figure, su I can suggest other ways to add legend labels?

VibhutiBansal-11 replied:
![image](https://user-images.githubusercontent.com/52831429/168469521-db8cd2ac-eaa3-4e2c-8013-e02bc595c914.png)

karm-patel replied:
Okay I'm looking into this, meanwhile can you please paste whole figure so I can give you final go-ahead to create a PR

VibhutiBansal-11 replied:
Here is the Gist
[Gist](https://gist.github.com/VibhutiBansal-11/a26ce056f3bb9c433baa9ed61bdafbf4)
![after](https://user-images.githubusercontent.com/52831429/168483866-03843002-f5b5-462c-9a75-445ec6a604a3.png)

karm-patel commented:
Thanks for posting the figure, Here I'm putting my workaround for subfigure-1, let me know how does it look?
![image](https://user-images.githubusercontent.com/59387624/168484714-9f7c1d00-d8bb-4e30-aa14-69805087bbb5.png)
So basically,
1. You don't need here `plt.gca().set_aspect('equal')`, since this is not a contour plot where we need equal scaling of x & y'
2. I think we can separate the legend symbols in this kind of cases

Code
```py
def plot_data(x,muTrue,savename,title,fig=None,ax=None):
    if ax is None:
        fig, ax = plt.subplots()
    plt1 = ax.plot(x[:, 0], x[:, 1], "o",  markerfacecolor="b",label='samples $(y_n)$')
    ax.set_ylim([-1, 1])
    ax.set_xlim([-1, 1])
    ax.set_title(title)
    ax.set_xlabel("$y_1$")
    ax.set_ylabel("$y_2$")
    plt2 = ax.plot(muTrue[0], muTrue[1], "x",  markerfacecolor="black", color="k",label='mean $(z)$')
    leg1 = ax.legend(handles=plt1,loc = "lower left",fontsize = 7, framealpha=0.5)
    ax.add_artist(leg1)
    leg2 = ax.legend(handles=plt2,loc = "upper left",fontsize = 7, framealpha=0.5)
    #plt.gca().set_aspect('equal')
    sns.despine()
    if len(savename) > 0:
        savefig(savename)
    return fig,ax
```

Once you do this you can create a PR on the main repo. but before that
1. set up `pre commit`, see [here](https://gist.github.com/karm-patel/655c3c6ba36d09b796e00d3cde79ad38) for more details.
2. Since this will be your first PR you should give your intro like Aadesh gave in his PR: https://github.com/probml/pyprobml/pull/818

karm-patel commented:
Small Change: 
1. Samples marker size was looking big, handles `marker_size` for latexify version and normal version using `is_latexify_enabled()` 
3. There was too much space between legend marker and legend text
![image](https://user-images.githubusercontent.com/59387624/168485337-354d8386-7f2b-4430-b4d8-bd1888cd7964.png)
Code:
```py
def plot_data(x,muTrue,savename,title,fig=None,ax=None):
    marker_size = 4 if is_latexify_enabled() else None
    if ax is None:
        fig, ax = plt.subplots()
    plt1 = ax.plot(x[:, 0], x[:, 1], "o",  markerfacecolor="b",label='samples $(y_n)$',markersize=marker_size )
    ax.set_ylim([-1, 1])
    ax.set_xlim([-1, 1])
    ax.set_title(title)
    ax.set_xlabel("$y_1$")
    ax.set_ylabel("$y_2$")
    plt2 = ax.plot(muTrue[0], muTrue[1], "x",  markerfacecolor="black", color="k",label='mean $(z)$', markersize=marker_size)
    leg1 = ax.legend(handles=plt1,loc = "lower left",fontsize = 8, framealpha=0.5, handletextpad=0.1)
    ax.add_artist(leg1)
    leg2 = ax.legend(handles=plt2,loc = "upper left",fontsize = 8, framealpha=0.5, handletextpad=0.1)
    #plt.gca().set_aspect('equal')
    sns.despine()
    if len(savename) > 0:
        savefig(savename)
    return fig,ax
```

VibhutiBansal-11 commented:
Here is the new image
![after](https://user-images.githubusercontent.com/52831429/168649879-7faecfae-6b40-45e2-86db-ca2f22d770ea.png)
Also I had a doubt [Link](https://github.com/probml/pyprobml/pull/818#issuecomment-1122688920) I have not jaxified multivariate_normal saw similar comment, should I make a PR with current code?

karm-patel commented:
@patel-zeel, Can you give your approval on this PR?

VibhutiBansal-11 commented on notebooks/book1/03/gauss_infer_2d.ipynb:
```diff
@@ -0,0 +1,752 @@
+{
```
Okay
